### ISSUE TITLE
Making Nil, False and True frozen introduces spec failures

### ISSUE THREAD

**Issue by tak1n** [rubinius/rubinius#3517]
This commit https://github.com/rubinius/rubinius/commit/f56eacf9a5f7587c08226df2816c173e3649c429 is making Nil, False and True to be frozen.

When running the 2.2 ci spec suite locally I get:

```
❯❯❯ bin/mspec ci
rubinius 2.5.8.c96 (2.1.0 c1f479d2 2015-10-10 3.5.0 JI) [x86_64-linux-gnu]
...................................................................................................................................................................................................................
...................................................................................................................................................................................................................
...................................................................................................................................................................................................................
...................................................................................................................................................................................................................
....................................................................................................................................................................E.............E................................
..................................................EE.E.............................................................................................................................................................
...................................................................................................................................................................................................................
...................................................................................................................................................................................................................
...................................................................................................................................................................................................................
...................................................................................................................................................................................................................
.............E...................................................

1)
Kernel#taint has no effect on immediate values ERROR
RuntimeError: can't modify frozen instance of NilClass
           { } in Object#__script__ at spec/ruby/core/kernel/taint_spec.rb:28
                         Array#each at kernel/bootstrap/array.rb:76
           { } in Object#__script__ at spec/ruby/core/kernel/taint_spec.rb:27
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/ruby/core/kernel/taint_spec.rb:4
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

2)
Kernel#untrusted? has no effect on immediate values ERROR
RuntimeError: can't modify frozen instance of NilClass
           { } in Object#__script__ at spec/ruby/core/kernel/untrusted_spec.rb:16
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/ruby/core/kernel/untrusted_spec.rb:4
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

3)
Module#attr_accessor allows creating an attr_accessor on an immediate class ERROR
RuntimeError: can't modify frozen instance of TrueClass
           { } in Object#__script__ at spec/ruby/core/module/attr_accessor_spec.rb:36
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/ruby/core/module/attr_accessor_spec.rb:4
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

4)
Module#attr_reader allows for adding an attr_reader to an immediate ERROR
RuntimeError: can't modify frozen instance of TrueClass
  Kernel(TrueClass)#instance_variable_set at kernel/common/kernel.rb:445
                 { } in Object#__script__ at spec/ruby/core/module/attr_reader_spec.rb:32
        BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
            { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                               Array#each at kernel/bootstrap/array.rb:76
                   Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                    Integer(Fixnum)#times at kernel/common/integer.rb:196
                               Array#each at kernel/bootstrap/array.rb:76
                        Object#__script__ at spec/ruby/core/module/attr_reader_spec.rb:4
                              Kernel.load at kernel/common/kernel.rb:498
        BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                      Array#each at kernel/bootstrap/array.rb:76
          Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
          Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
                  Rubinius::Loader#script at kernel/loader.rb:656
                    Rubinius::Loader#main at kernel/loader.rb:842

5)
Module#attr_writer allows for adding an attr_writer to an immediate ERROR
RuntimeError: can't modify frozen instance of TrueClass
           { } in Object#__script__ at spec/ruby/core/module/attr_writer_spec.rb:32
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/ruby/core/module/attr_writer_spec.rb:4
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

6)
Rubinius::Type.infect is a no-op for true ERROR
RuntimeError: can't modify frozen instance of TrueClass
           { } in Object#__script__ at spec/core/type/infect_spec.rb:51
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/core/type/infect_spec.rb:3
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

7)
Rubinius::Type.infect is a no-op for false ERROR
RuntimeError: can't modify frozen instance of FalseClass
           { } in Object#__script__ at spec/core/type/infect_spec.rb:55
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/core/type/infect_spec.rb:3
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

8)
Rubinius::Type.infect is a no-op for nil ERROR
RuntimeError: can't modify frozen instance of NilClass
           { } in Object#__script__ at spec/core/type/infect_spec.rb:59
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
      { } in Enumerable(Array)#all? at kernel/common/enumerable.rb:389
                         Array#each at kernel/bootstrap/array.rb:76
             Enumerable(Array)#all? at kernel/common/enumerable.rb:389
              Integer(Fixnum)#times at kernel/common/integer.rb:196
                         Array#each at kernel/bootstrap/array.rb:76
                  Object#__script__ at spec/core/type/infect_spec.rb:3
                        Kernel.load at kernel/common/kernel.rb:498
  BasicObject(Object)#instance_eval at kernel/common/eval.rb:43
                         Array#each at kernel/bootstrap/array.rb:76
   Rubinius::CodeLoader#load_script at kernel/delta/code_loader.rb:66
   Rubinius::CodeLoader.load_script at kernel/delta/code_loader.rb:152
            Rubinius::Loader#script at kernel/loader.rb:656
              Rubinius::Loader#main at kernel/loader.rb:842

Finished in 132.184115 seconds

2174 files, 20522 examples, 154546 expectations, 0 failures, 8 errors
```

**Comment by YorickPeterse** [comment]
Looking at the specs (https://github.com/rubinius/rubinius/blob/c1f479d2b4ed707fbe942c64f1a63693fa338290/spec/ruby/core/kernel/taint_spec.rb#L27-L29) they literally test tainting these values. MRI seems to ignore `taint` when calling it on nil/true/false:

    true.taint # => true
    true.tainted? # => false

We can probably solve this by just adding a simple check, something like:

    if true == self or false == self or nil == self
      return self
    end

    ... everything else ...

**Comment by chuckremes** [comment]
Solved by a2d76f56c3cec.
